README.md
Marco Cacciabue, Débora N. Marcone

<!-- README.md is generated from README.Rmd. Please edit that file -->

# **INFINITy** <img src='man/figures/hex.png' style="float:right; height:200px;" /> a fast and accurate tool for classifying human influenza A and B virus sequences

<!-- badges: start -->

[![DOI](https://zenodo.org/badge/COMPLETE!!)](https://zenodo.org/badge/latestdoi/COMPLETE!!)
[![](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
[![R-CMD-check](https://github.com/marcocacciabue/infinity/workflows/R-CMD-check/badge.svg)](https://github.com/marcocacciabue/infinity/actions)
<!-- badges: end -->

If you wish to run the app without installation and directly on the web
go to one the following servers:

## [INFINITy web-application official server!](https://infinity.unlu.edu.ar/)

## [INFINITy web-application secondary server!](https://cacciabue.shinyapps.io/infinity2/)

If you find our app useful please consider citing: [Cacciabue et al
2023](https://doi.org/10.1111/irv.13096)

## :arrow_double_down: Installation

If you have a working R and Rstudio setup you can install the released
version of **INFINITy** from [GitHub](https://github.com/) with:

``` r
if (!require("remotes", quietly = TRUE))
  install.packages("remotes")
  
remotes::install_github("marcocacciabue/infinity")
```

This could take several minutes depending on your system and
installation. Only the first time it runs.

Alternatively, you can download the repository as a .zip file and
install it manually with Rstudio.

## :computer: Deploying **INFINITy**

To start the app simply run the following command:

``` r
infinity::runShinyApp()
```

:+1: You are ready to classify your data.

## **INFINITy** is an R package

This means that users that prefer working directly in an R console can
use some of the exported functions. The easiest way is to use the
wrapper funtion “ClassifyInfinity()”

``` r
#load the library
library(infinity)

# Indicate the file path to the fasta file to use.If your file is in your working directory you need to simply indicate the file name. In this case, we use a test file provided with the package itself. 

file_path<-system.file("extdata","test_dataset.fasta",package="infinity")

# Use the wrapper function. You can change the classification model and pass other arguments
ClassifyInfinity(inputFile=file_path,model=FULL_HA)

# This command run the whole pipeline and saves a file in the working directory a "Results.csv" by default. You can change the name file setting the "outputFile" parameter.
```

For a more detailed workflow the user can use the exported functions of
the package.

``` r
# First we indicate the location of the fasta file. In this case, we use a test file provided with the package itself.
file_path<-system.file("extdata","test_dataset.fasta",package="infinity")

# We load the sequences
sequence<-ape::read.FASTA(file_path,type = "DNA")

# We the count and normalize the k-mers
NormalizedData <- Kcounter(SequenceData=sequence,model=FULL_HA)

# We perform the classification
PredictedData <- PredictionCaller(NormalizedData=NormalizedData,model=FULL_HA)

# We add the quality FLAGs
PredictedData <- QualityControl(PredictedData,model=FULL_HA)

# We adjust the classification according to the FLAGs present for each sample:
PredictedData <- Quality_filter(PredictedData)

# The PredictedData dataframe contains the classifications

PredictedData
```

## :whale: Docker images available

### **INFINITy** as a shiny app in docker

Another way to run **INFINITy** as a shiny app is to use the docker
image. Follow these steps:

1.  If you don´t already have it, install docker:
    <https://www.docker.com/get-started>.

2.  Open a terminal and run the following:

``` bash
docker pull cacciabue/infinity:shiny
```

and wait for the image to download. You only have to run this command
the first time, or whenever you want to check for updates.

3.  Once the downloading is complete run the following:

``` bash
docker run -d --rm -p 3838:3838 cacciabue/infinity:shiny
```

4.  Finally, open your favorite browser and go to
    <http://localhost:3838/>
5.  The app should be up and running in your browser. Load the fasta
    file and press RUN.
6.  You can save a report using the corresponding button.

### **INFINITy** in a docker image with all dependecies allready installed

For more reproducibility a fully operational environment is available to
work directly in docker:

1.  If you don´t already have it, install docker:
    <https://www.docker.com/get-started>.

2.  Open a terminal, go to the directory where the fasta files are
    stored and run the following:

``` bash
docker pull cacciabue/infinity:cli
```

3.  Once the downloading is complete run the following:

``` bash
#for WINDOWS 
docker run --rm --volume %cd%:/nexus cacciabue/infinity:cli

#for unix/MAC
docker run -it --volume $(pwd):/nexus cacciabue/infinity:cli
```

4.  Now you can run

``` bash
setwd('nexus')

# Indicate the file path to the fasta file to use.If your file is in your working directory you need to simply indicate the file name. In this case, we use a test file provided with the package itself. 

file_path<-system.file("extdata","test_dataset.fasta",package="infinity")

# Use the wrapper function. You can change the classification model and pass other arguments
ClassifyInfinity(inputFile=file_path,model=FULL_HA)

# This command run the whole pipeline and saves a file in the working directory a "Results.csv" by default. You can change the name file setting the "outputFile" parameter.

# To exit the container just run

q()
```

5.  Alternatively you can perform step 3 and 4 in one command like this:

``` bash

#for WINDOWS 

docker run --rm --volume %cd%:/nexus cacciabue/infinity:cli R -e "setwd('nexus');library('infinity');ClassifyInfinity(inputFile='test_dataset.fasta',model=FULL_HA)"

#for unix/MAC

docker run --rm --volume $(pwd):/nexus cacciabue/infinity:cli R -e "setwd('nexus');library('infinity');ClassifyInfinity(inputFile='test_dataset.fasta',model=FULL_HA)"


# USER SHOULD CHANGE test_dataset.fasta for the correct file name
```


NEWS.md
# infinity 0.3.2
* Added correct CITATION information. You can get this info by running citation("infinity")
# infinity 0.3.1
* Changed Clade definition 3C.2a1b+T135N to 3C.2a1b.3.
* minor typos fixed.
# infinity 0.3.0
* added support for running infinity directly on a R console.
* updated README file.
# infinity 0.2.0

# infinity 0.1.0

# infinity 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(infinity)

test_check("infinity")


tests/testthat/test_PredictionCaller.R
# library("ranger")
library("infinity")


file_path<-system.file("extdata","test_dataset.fasta",package="infinity")

sequence<-ape::read.FASTA(file_path,type = "DNA")

NormalizedData<-Kcounter(SequenceData=sequence,model=FULL_HA)

calling_null<-ranger::predictions(FULL_HA)


PredictedData <- PredictionCaller(NormalizedData=NormalizedData,model=FULL_HA)


test_that("A dataframe is produced with the corresponding results ", {
  expect_true(is.data.frame(PredictedData))
  expect_true(length(colnames(PredictedData))==8)
  expect_true("Label" %in% colnames(PredictedData))

})

test_that("The number of results in dataframe is equal to number of test sequences", {
  expect_true(length(sequence)==length(PredictedData$Label))

})



# NormalizedData<-Kcounter(SequenceData=sequence,model=HA1)
#
# calling_null<-ranger::predictions(HA1)
#
#
# PredictedData <- PredictionCaller(NormalizedData=NormalizedData,model=HA1)
# PredictedData
# PredictedData <- QualityControl(model=FULL_HA,PredictedData)
# Stringent_filter(PredictedData)
# Quality_filter(PredictedData)

test_that("A dataframe is produced with the corresponding results ", {
  expect_true(is.data.frame(PredictedData))
  expect_true(length(colnames(PredictedData))==8)
  expect_true("Label" %in% colnames(PredictedData))

})

test_that("The number of results in dataframe is equal to number of test sequences", {
  expect_true(length(sequence)==length(PredictedData$Label))

})



R/data.R
#' Random forest classification model based on full length HA sequence
#'
#' This a classification model (created with the ranger package) to use with the infinity package
#'
#'
#' @format The model includes a list with 18 elements (modified from a Object of class `ranger`):
#' \describe{
#'   \item{predictions}{training dataset predictions}
#'   \item{num.trees}{Number of decision trees (ussually 1000)}
#'   \item{num.independent.variables}{Number of independent variables}
#'   \item{mtry}{Value of mtry used}
#'   \item{min.node.size}{Value of minimal node size used.}
#'   \item{prediction.error}{Overall out of bag prediction error.}
#'   \item{forest}{Saved forest}
#'   \item{confusion.matrix}{Contingency table for classes and predictions based on out of bag samples }
#'   \item{splitrule}{tSplit rule used for training}
#'   \item{num.random.splits}{Number of random splits}
#'   \item{treetype}{Type of forest/tree}
#'   \item{call}{Function call}
#'   \item{importance.mode}{Importance mode used}
#'   \item{num.samples}{Number of samples}
#'   \item{kmer}{k-mer size used for preprocessing}
#'   \item{info}{extra information of the classification model (optional)}
#'   \item{date}{date of model creation}
#' }
"FULL_HA"

#' Random forest classification model based on partial length HA1 sequence
#'
#' This a classification model (created with the ranger package) to use with the infinity package
#'
#'
#' @format The model includes a list with 18 elements (modified from a Object of class `ranger`):
#' \describe{
#'   \item{predictions}{training dataset predictions}
#'   \item{num.trees}{Number of decision trees (ussually 1000)}
#'   \item{num.independent.variables}{Number of independent variables}
#'   \item{mtry}{Value of mtry used}
#'   \item{min.node.size}{Value of minimal node size used.}
#'   \item{prediction.error}{Overall out of bag prediction error.}
#'   \item{forest}{Saved forest}
#'   \item{confusion.matrix}{Contingency table for classes and predictions based on out of bag samples }
#'   \item{splitrule}{tSplit rule used for training}
#'   \item{num.random.splits}{Number of random splits}
#'   \item{treetype}{Type of forest/tree}
#'   \item{call}{Function call}
#'   \item{importance.mode}{Importance mode used}
#'   \item{num.samples}{Number of samples}
#'   \item{kmer}{k-mer size used for preprocessing}
#'   \item{info}{extra information of the classification model (optional)}
#'   \item{date}{date of model creation}
#' }
"HA1"


R/utils.R


#' ModelControl
#' Utility function to check the classification model object
#'
#' @param model  random forest classification model. Must be FULL_HA or HA1. See [FULL_HA] and [HA1].
#'
#' @return warnings
#'
#' @export
#'
#' @examples
#' ModelControl(FULL_HA)
ModelControl<- function(model){

  if (is.null(model)){
    stop("Model argument must be indicated and cannot be null. Try model=FULL_HA or model=HA1")
  }

  if (methods::is(model)!="ranger"){
    stop("Model must be a ranger object. Try model=FULL_HA or model=HA1")
  }

  if (!"info" %in% names(model)){
    stop("Model must include the info object. Try model=FULL_HA or model=HA1")
  }
  if (!"date" %in% names(model)){
    stop("Model must include the date object. Try model=FULL_HA or model=HA1")
  }


}


#' Ignore unused imports
#'
#' @return nothing
#'
#'
ignore_unused_imports <- function() {
  # elminate NOTE y R CMD CHECK as suggested in https://r-pkgs.org/dependencies-in-practice.html
  DT::dataTableOutput
  bslib::bootstrap
  shinyWidgets::actionGroupButtons
  shinyjs::click
  tibble::add_row
  rmarkdown::clean_site
  }


R/ClassifyInfinity.R

#' Infinity pipeline
#'
#' Wrapper function that performs all the recommended step. This includes
#' \describe{
#'   \item{Kcounter}{# Count Kmers and normalize data}
#'   \item{PredictionCaller}{Run prediction pipeline}
#'   \item{QualityControl}{Check Quality of the input data and classification results}
#'   \item{Quality_filter}{For sequences with any FLAG present set result Clade to LowQuality}
#' }
#'
#' @param inputFile string path relative to the working directory of the input file. Must be in fasta format.
#' @param outputFile string path relative to the working directory of the output file. Default = "results.csv"
#' @inheritParams QualityControl
#'
#' @return a comma-delimited file
#' @export
#'
#' @examples
#'
#' \dontrun{
#' file_path<-system.file("extdata","test_dataset.fasta",package="infinity")
#'
#' ClassifyInfinity(inputFile=file_path,model=FULL_HA)
#' }
ClassifyInfinity<-function(inputFile,
                           outputFile="results.csv",
                           model,
                           QC_value=0.4,
                           Length_value=0.5,
                           N_value=2){

  sequence<-ape::read.FASTA(inputFile,type = "DNA")

  # Count Kmers and normalize data

  NormalizedData <- Kcounter(SequenceData=sequence,model=model)
  #  Run prediction pipeline

  PredictedData <- PredictionCaller(NormalizedData=NormalizedData,model=model)

  #  Check Quality of the input data and classification results
  PredictedData <- QualityControl(PredictedData,
                                  model=model,
                                  QC_value=QC_value,
                                  Length_value=Length_value,
                                  N_value=N_value)
  # For sequences with any FLAG present set result Clade to LowQuality

  PredictedData <- Quality_filter(PredictedData)

  utils::write.csv2(PredictedData,outputFile)

}


R/PredictionCaller.R

#' PredictionCaller
#'
#' Performs the prediction and computes probability values.
#'
#' @param NormalizedData A list of 3 elements: normalized k-mer counts, genome length and contents of undefined bases.Produced by the [Kcounter()] function
#' @inheritParams Kcounter
#' @param QC_unknown numeric value from 0 to 1. Stringent filter, do not classify below this probability score (default = 0.2)
#' @return Data.frame with the classification results and quality checks.
#' The output has the following properties:
#' * Each row corresponds to one sample sequence.
#' * `Label` is the name of the sequence.
#' * `Clade` is the corresponding prediction.
#' * `Probability` is the proportions of trees that agreed with the Clade result. Values between 0 to 1.
#' * `Probability_QC` a logical value. If `TRUE` the sequence passed the quality filter for probability. Run [QualityControl()] to fill this column.
#' * `Length` Sequence length.
#' * `Length_QC` a logical value. If `TRUE` the sequence passed the quality filter for length.Run [QualityControl()] to fill this column
#' * `N` proportions of undefined bases in the sequence. The lower the better.
#' * `N_QC` a logical value. If `TRUE` the sequence passed the quality filter for undefined bases. Run [QualityControl()] to fill this column
#'
#' @export
#' @importFrom ranger predictions
#' @examples
#'
#'
#'
#' file_path<-system.file("extdata","test_dataset.fasta",package="infinity")
#'
#' sequence<-ape::read.FASTA(file_path,type = "DNA")
#'
#' NormalizedData <- Kcounter(SequenceData=sequence,model=FULL_HA)
#'
#' PredictedData <- PredictionCaller(NormalizedData=NormalizedData,model=FULL_HA)
#'

PredictionCaller<-function(NormalizedData,
                           model,
                           QC_unknown=0.2){

   if (is.null(NormalizedData) | missing(NormalizedData)){
    stop("'NormalizedData' must be indicated")
  }

  ModelControl(model)

  calling<-predict(model,
                               NormalizedData$DataCount)
  #Run the predict method from de Ranger package, retaining the classification result from each tree in the model (to calculate a probability value for each classification)
  calling_all<-predict(model,
                                   NormalizedData$DataCount,
                                   predict.all = TRUE)
  probability <- rep(0, length(calling_all$predictions[,1]))

  for (i in 1:length(calling_all$predictions[,1])) {
    #extract predictions for each SequenceData sample in temp vector,
    #count the number of correct predictions and divide by number of trees to get a probability.
    temp<-calling_all$predictions[i,]
    probability[i] <- sum(temp==which(model$forest$levels==calling$predictions[i]))/model$num.trees

  }


  # QualityList<-QualityControl(n_length=NormalizedData$n_length,
  #                          genome_length=NormalizedData$genome_length,
  #                          probability,
  #                          model)
  samples<-rep(0,length(row.names(NormalizedData$DataCount)))
  data<-data.frame(Label= row.names(NormalizedData$DataCount),
             Clade=calling$prediction,
             Probability=probability,
             Length=NormalizedData$genome_length,
             Length_QC=samples,
             N=NormalizedData$n_length,
             N_QC=samples,
             Probability_QC=samples)


  filter<-data$Probability<QC_unknown
  data$Clade <- as.character(data$Clade)
  data$Clade[filter]<-"unknown"


  return(data)
}



#' QualityControl
#'
#' @param data data.frame obtained with [PredictionCaller()]
#' @param QC_value numeric value from 0 to 1. (default = 0.4)
#' @param Length_value numeric value from 0 to 1. Proportion of difference to the expected sequence length. (default = 0.5)
#' @param N_value numeric value from 0 to 100. Percentage of acceptable ambiguous bases. (default = 2)
#' @inheritParams Kcounter
#' @return A list with three logical vectors. In each case TRUE means pass.
#'
#' @export
#'
#' @examples
#'
#' file_path<-system.file("extdata","test_dataset.fasta",package="infinity")
#'
#' sequence<-ape::read.FASTA(file_path,type = "DNA")
#'
#' NormalizedData <- Kcounter(SequenceData=sequence,model=FULL_HA)
#'
#' PredictedData <- PredictionCaller(NormalizedData=NormalizedData,model=FULL_HA)
#'
#' QualityControl(PredictedData,model=FULL_HA)
#'
#'
QualityControl<-function(data,
                         QC_value=0.4,
                         Length_value=0.5,
                         N_value=2,
                         model){
  ModelControl(model)

  if(!is.numeric(QC_value)){
    stop("`QC_value` must be numeric")
  }
  if(!is.numeric(Length_value)){
    stop("`Length_value` must be numeric")
  }
  if(!is.numeric(N_value)){
    stop("`N_value` must be numeric")
  }

  if(QC_value>1 | QC_value<0.2){
    stop("`QC_value` must be numeric and between 0.2 to 1")
  }

  if(N_value>10 | N_value<0){
    stop("`N_value` must be numeric and between 0 to 10")
  }

  if(Length_value>1 | Length_value<0){
    stop("`Length_value` must be numeric and between 0 to 1")
  }

  data$N_QC<-(data$N<N_value)

  if ("Flu"==model$info){
    data$Length_QC<-(data$Length>(1-Length_value)*1700)&(data$Length<(1+Length_value)*1700)
  }
  if ("Flu_Ha1"==model$info){
    data$Length_QC<-(data$Length>(1-Length_value)*1100)&(data$Length<(1+Length_value)*1100)
  }
  data$Probability_QC<-data$Probability>QC_value


  return(data)

}








#' Quality_filter
#'
#' Change Clade definition to "LowQuality" in samples with at least one FLAG from [QualityControl()].
#'
#' @param data data.frame obtained with [QualityControl()]
#'
#' @return data.frame
#' @export
#'
#' @examples
#'
#' file_path<-system.file("extdata","test_dataset.fasta",package="infinity")
#'
#' sequence<-ape::read.FASTA(file_path,type = "DNA")
#'
#' NormalizedData <- Kcounter(SequenceData=sequence,model=FULL_HA)
#'
#' PredictedData <- PredictionCaller(NormalizedData=NormalizedData,model=FULL_HA)
#'
#' PredictedData <- QualityControl(PredictedData,model=FULL_HA)
#'
#' PredictedData <- Quality_filter(PredictedData)
#'
Quality_filter<- function(data){

  filter<-!(data$Length_QC & data$N_QC & data$Probability_QC)
  filter<- filter & (data$Clade!="unknown")
  data$Clade <- as.character(data$Clade)
  data$Clade[filter]<-"LowQuality"


  return(data)
}




R/CounterNormalizer.R

#' Kcounter
#'
#' Counts k-mers of the size required by the input model and normalize the data regarding genome size.
#'
#'
#' @param SequenceData A list of DNA sequences. Object must be created with the `ape` package. See example.
#' @param model A random forest classification model. Must be FULL_HA or HA1. See [FULL_HA] and [HA1].
#'
#' @return A list of 3 vectors: normalized k-mer counts, genome length and contents of undefined bases.
#'
#' @export
#'
#' @examples
#'
#' file_path<-system.file("extdata","test_dataset.fasta",package="infinity")
#'
#' sequence<-ape::read.FASTA(file_path,type = "DNA")
#'
#' NormalizedData<-Kcounter(SequenceData=sequence,model=FULL_HA)

Kcounter<-function(SequenceData,
                    model){
  ModelControl(model)
  DataCount<-kmer::kcount(SequenceData , k=model$kmer)
  genome_length<-0
  n_length<-0
  for(i in 1:length(DataCount[,1])){

    k<-SequenceData[i]
    k<-as.matrix(k)
    DataCount[i,]<- DataCount[i,]*model$kmer/(length(k))
    genome_length[i]<-length(k)
    n_length[i]<-round(100*ape::base.freq(k,all = TRUE)[15],2)
  }
 return(list(DataCount=DataCount,
             genome_length=genome_length,
             n_length=n_length))
}





inst/myapp/app.R
# library(renv)
library(shiny)
library(bslib)
# library(ranger)
# library(kmer)
# library(ape)
library(DT)
# library(seqinr)

options(shiny.maxRequestSize = 2000*1024^2)
# library(rintrojs)
# library(magrittr)
# library(dplyr)
library(shinyjs)
#for mac, without this Rstudio crashes
Sys.setenv(LIBGL_ALWAYS_SOFTWARE=1)
# library(tibble)
# library(RColorBrewer)
# library(parallel)
# library(shinylogs)
library(shinyWidgets)
# library(RColorBrewer)


theme_INFINITy <- bslib::bs_theme(bg = "#FFFFFF",
                       fg = "#4e79cb",
                       primary = "#ffc72c",
                       secondary= "#ee65cd",
                       font_scale = 0.8,
                       base_font = font_google("Prompt"))

ui<-shinyUI(


  navbarPage(title = span(img(src="logo_a.png", style="margin-top: -5px;", height = 50),
                          "INFINITy"),
             # add div("Enabled by data from",tags$a(img(style="width: 50px", src = "GISAID.png"),href="https://www.gisaid.org/"))
             theme= theme_INFINITy,
             tags$head(
               tags$link(rel = "icon", type = "image/png", sizes = "32x32", href = "logo_a.png")),
             windowTitle ="INFINITy",
             selected = "Home",
             footer = list(column(12, align="center",
                                  HTML("<br></br>"),
                             strong("GISAID data provided on this website are subject to GISAID's"),
                             tags$a(href="https://www.gisaid.org/DAA/",strong("Terms and Conditions"))),
                           column(12, align="center", class = "footer",
                               includeHTML("www/footer_infinity.html"))),
    tabPanel("Home",
      fluidPage(align="center",
        includeHTML("www/title.html"),
        div("Enabled by data from",tags$a(img(style="width: 50px", src = "GISAID.png"),href="https://www.gisaid.org/")),
        br(),
        textAreaInput('SequenceData', 'Paste your sequences in FASTA format into the field below', value = "", placeholder = "", width = "70%", height="100px"),
        fileInput("file", "Submit a file with your sequences in FASTA format",
                  accept = c(".text",".fasta",".fas",".fasta")),
        radioGroupButtons(
          inputId = "select",
          label = "Choose the model according your SequenceData length sequences",
          choices = c("FULL_HA", "HA1"),
          status = "primary"),
        checkboxInput("advanceOptions", "Advanced options", FALSE),
        conditionalPanel(condition='input.advanceOptions',
        checkboxInput("qualityfilter", "Classify even if quality of sequence is low? (Not recommended)", FALSE),
        sliderInput("QC_value", "Probability threshold (default 0.4):",
                    min = 0.2, max = 1,
                    value = 0.4, step = 0.05),
        sliderInput("N_value", "Percentage of acceptable ambiguous bases (default 2):",
                    min = 0.1, max = 10,
                    value = 2, step = 0.05),
        sliderInput("Length_value", "Proportion of difference to the expected sequence length (default 0.5): :",
                    min = 0.1, max = 1,
                    value = 0.5, step = 0.05)),
        br(),
        actionButton("go", "RUN"),
        HTML("<br><br><br>"),
        br(),
        conditionalPanel(
          condition='output.table!=null && output.table!="A"',
          div(strong("The following sequences have passed the quality test"))),
        br(),
        DT::dataTableOutput("table"),
        br(),
        conditionalPanel(
          condition='output.table_reject!=null && output.table_reject!=""',
          div(strong("The following sequences have"),strong("NOT passed",style="color:red"),strong(" the quality test"))),
        br(),
        DT::dataTableOutput("table_reject"),
        conditionalPanel(
          condition='output.table!=null && output.table!=""',
          downloadButton('report',"Generate report (ENG)")))),
    tabPanel("Instructions",
             align="center",
             includeHTML("www/title.html"),
             div("Enabled by data from",tags$a(img(style="width: 50px", src = "GISAID.png"),href="https://www.gisaid.org/")),
             includeHTML("www/instructions.html")),
    tabPanel("Classification model information",DT::dataTableOutput("table2"))
  ))

server <- function(input, output, session) {
  values <- reactiveValues(SequenceData_FILE = NULL)
  observeEvent(input$go, {
    if(length(input$file)==0&(input$SequenceData=="")){
      showModal(modalDialog(
        title = "Important message", easyClose = TRUE,
        "Please load the fasta file (or the sequence) first and then press RUN.
Also, remember that the file must NOT exceed 2 MB in size.
"
      ))}})
  # model_reactive <- eventReactive(input$select,{
  #
  #   model <- readRDS(paste("models/",sep="",input$select,".rds"))
  #   list(model=model)
  # })
  model_reactive <- eventReactive(input$select,{
  if (input$select=="FULL_HA"){
    model <- infinity::FULL_HA}else{
    model <- infinity::HA1
    }
    list(model=model)
  })
  SequenceData_data<- observeEvent(input$SequenceData,{
    req(input$SequenceData)
    #gather input and set up temp file
    SequenceData_tmp <- input$SequenceData
    tmp <- tempfile(fileext = ".fa")


    #this makes sure the fasta is formatted properly
    if (startsWith(SequenceData_tmp, ">")){
      writeLines(SequenceData_tmp, tmp)
    } else {
      writeLines(paste0(">SequenceData\n",SequenceData_tmp), tmp)
    }
    values$SequenceData_FILE<-tmp
  })


  SequenceData_data<- observeEvent(input$file,{
    req(input$file)

    values$SequenceData_FILE<-input$file[1,4]

  })



  data_reactive<- eventReactive(input$go,{
    req(values$SequenceData_FILE)
    progress <- shiny::Progress$new()
    on.exit(progress$close())
    progress$set(message = "processing data", value = 0)
    progress$inc(0.2, detail = paste("Reading fasta"))



    tmp<-values$SequenceData_FILE
    SequenceData<-ape::read.FASTA(tmp,type = "DNA")

    model <- model_reactive()$model


    #Calculate k-mer counts from SequenceData sequences
    progress$inc(0.4, detail = paste("Counting kmers"))

    NormalizeData<-infinity::Kcounter(SequenceData,
               model)


    data_out <-  infinity::PredictionCaller(NormalizeData,
                                  model=model)
    # data_out<-infinity::QualityControl(model=model,
    #                                    data=data_out)
    list(message="Done!",data_out=data_out)
  })

  data_predicted<-reactive({
    # req(values$SequenceData_FILE)
    model <- model_reactive()$model
    data_out<-data_reactive()$data_out

    data_out<-infinity::QualityControl(model=model,
                                       data=data_out,
                                       QC_value=input$QC_value,
                                       N_value=input$N_value,
                                       Length_value=input$Length_value)
    if(input$qualityfilter==FALSE){
    data_out_filtered<-infinity::Quality_filter(data=data_out)
    }else{
    data_out_filtered<-data_out
    }

    data_out_filtered
  })

  output$text <- renderText({
    data_reactive()$message
  })

  table<-reactive({

    data_out<-data_predicted()

    data_out

  })

  table_pass<-reactive({

    table<-table()
    # table<-table %>% filter(Probability_QC == 1 & N_QC == 1 & Length_QC == 1)
    filter<-(table$Probability_QC == 1 & table$N_QC == 1 & table$Length_QC == 1)
    table<-table[filter,]

    # table<-table[table$Probability_QC == 1,]
  })


  table_reject<-reactive({

    table<-table()
    # table<-table %>% filter(Probability_QC == 0 | N_QC == 0 | Length_QC == 0)

    # table<-table[table$Probability_QC == 0,]
    filter<-(table$Probability_QC == 0 | table$N_QC == 0 | table$Length_QC == 0)
    table<-table[filter,]

  })

  output$table <- DT::renderDataTable({

    col2<-"#ee65cd"
    col<-"#ffc72c"

    table<-table_pass()


    datatable(table,selection = 'single',
              style = 'bootstrap',
              extensions = 'Buttons',
              options = list(
                columnDefs = list(list(targets = c(5,7,8), visible = FALSE)),
                dom = 'Bfrtip',
                lengthMenu = list(c(5, 15, -1), c('5', '15', 'All')),
                pageLength = 15,
                buttons =
                  list('copy', 'print', list(
                    extend = 'collection',
                    buttons = list(
                      list(extend = 'csv', filename = paste(input$file[1,1],"INFINITy_results"),sep=""),
                      list(extend = 'excel', filename = paste(input$file[1,1],"INFINITy_results"),sep=""),
                      list(extend = 'pdf', filename = paste(input$file[1,1],"INFINITy_results"),sep="")),
                    text = 'Download'
                  ),
                list(
                  extend = "collection",
                  text = 'Show All',
                  action = DT::JS("function ( e, dt, node, config ) {
                                    dt.page.len(-1);
                                    dt.ajax.reload();
                                }")
                ))
              ))%>%formatStyle("Length","Length_QC",backgroundColor = styleEqual(c(0, 1), c(col2, col)))%>%
                   formatStyle("N","N_QC",backgroundColor = styleEqual(c(0, 1), c(col2, col)))%>%
                   formatStyle("Probability","Probability_QC",backgroundColor = styleEqual(c(0, 1), c(col2, col)))


    })


  output$table_reject <- DT::renderDataTable({

    col2<-"#ee65cd"
    col<-"#ffc72c"

    table<-table_reject()


    datatable(table,selection = 'single',
              style = 'bootstrap',
              extensions = 'Buttons',
              options = list(
                columnDefs = list(list(targets = c(5,7,8), visible = FALSE)),
                dom = 'Bfrtip',
                lengthMenu = list(c(5, 15, -1), c('5', '15', 'All')),
                pageLength = 15,
                buttons =
                  list('copy', 'print', list(
                    extend = 'collection',
                    buttons = list(
                      list(extend = 'csv', filename = paste(input$file[1,1],"INFINITy_results"),sep=""),
                      list(extend = 'excel', filename = paste(input$file[1,1],"INFINITy_results"),sep=""),
                      list(extend = 'pdf', filename = paste(input$file[1,1],"INFINITy_results"),sep="")),
                    text = 'Download'
                  ),
                  list(
                    extend = "collection",
                    text = 'Show All',
                    action = DT::JS("function ( e, dt, node, config ) {
                                    dt.page.len(-1);
                                    dt.ajax.reload();
                                }")
                  ))
              ))%>%formatStyle("Length","Length_QC",backgroundColor = styleEqual(c(0, 1), c(col2, col)))%>%
      formatStyle("N","N_QC",backgroundColor = styleEqual(c(0, 1), c(col2, col)))%>%
      formatStyle("Probability","Probability_QC",backgroundColor = styleEqual(c(0, 1), c(col2, col)))
        })

  model_table<-reactive({
    model1<-model_reactive()$model
    model1_data<-data.frame(Model=model1$info,date=model1$date,trees=model1$num.trees,Oob= round(model1$prediction.error,4))

    model1_data
  })

  output$table2 <- DT::renderDataTable({
    datatable(model_table())


  })


  output$report <- downloadHandler(
    # For PDF output, change this to "report.pdf"
    filename = "report.html",
    content = function(file) {
      # Copy the report file to a temporary directory before processing it, in
      # case we don't have write permissions to the current working dir (which
      # can happen when deployed).
      tempReport <- file.path(tempdir(), "report.rmd")
      file.copy("report.rmd", tempReport, overwrite = TRUE)

      # Set up parameters to pass to Rmd document

      # Knit the document, passing in the `params` list, and eval it in a
      # child of the global environment (this isolates the code in the document
      # from the code in this app).
      rmarkdown::render(tempReport, output_file = file)

    }
  )

}

shinyApp(ui = ui, server = server)

